pySDC/implementations/problem_classes/AdvectionEquation_ND_FD.py

```diff
@@ -48,91 +48,84 @@ def __init__(
+            raise ProblemError('need a square domain, got %s' % nvars)
 
         # invoke super init, passing number of dofs, dtype_u and dtype_f
-        super(advectionNd, self).__init__(
+        super().__init__(
             init=(nvars, None, np.dtype('float64')),
             dtype_u=mesh,
             dtype_f=mesh,
-            nvars=nvars,
-            c=c,
-            freq=freq,
-            stencil_type=stencil_type,
-            order=order,
-            lintol=lintol,
-            liniter=liniter,
-            direct_solver=direct_solver,
-            bc=bc,
-            ndim=ndim,
         )
 
-        if self.params.ndim == 1:
-            if type(self.params.nvars) is not tuple:
-                self.params.nvars = (self.params.nvars,)
-            if type(self.params.freq) is not tuple:
-                self.params.freq = (self.params.freq,)
-
         # compute dx (equal in both dimensions) and get discretization matrix A
-        if self.params.bc == 'periodic':
-            self.dx = 1.0 / self.params.nvars[0]
-            xvalues = np.array([i * self.dx for i in range(self.params.nvars[0])])
-        elif self.params.bc == 'dirichlet-zero':
-            self.dx = 1.0 / (self.params.nvars[0] + 1)
-            xvalues = np.array([(i + 1) * self.dx for i in range(self.params.nvars[0])])
+        if bc == 'periodic':
+            xvalues = np.linspace(0, 1, num=nvars[0], endpoint=False)
+        elif bc == 'dirichlet-zero':
+            xvalues = np.linspace(0, 1, num=nvars[0] + 2)[1:-1]
         else:
             raise ProblemError(f'Boundary conditions {self.params.bc} not implemented.')
+        dx = xvalues[1] - xvalues[0]
 
         self.A = problem_helper.get_finite_difference_matrix(
             derivative=1,
-            order=self.params.order,
-            stencil_type=self.params.stencil_type,
-            dx=self.dx,
-            size=self.params.nvars[0],
-            dim=self.params.ndim,
-            bc=self.params.bc,
+            order=order,
+            stencil_type=stencil_type,
+            dx=dx,
+            size=nvars[0],
+            dim=ndim,
+            bc=bc,
         )
-        self.A *= -self.params.c
+        self.A *= -c
+
+        self.xvalues = xvalues
+        self.Id = sp.eye(np.prod(nvars), format='csc')
 
-        self.xv = np.meshgrid(*[xvalues for _ in range(self.params.ndim)])
-        self.Id = sp.eye(np.prod(self.params.nvars), format='csc')
+        # store relevant attributes
+        self.nvars, self.ndim, self.c = nvars, ndim, c
+        self.stencil_type, self.order, self.bc = stencil_type, order, bc
+        self.freq, self.sigma = freq, sigma
+        self.lintol, self.liniter, self.direct_solver = lintol, liniter, direct_solver
+
+        # register parameters
+        self._register('nvars', 'ndim', 'c', 'stencil_type', 'order', 'bc', readOnly=True)
+        self._register('freq', 'sigma', 'lintol', 'liniter', 'direct_solver')
```

Comment:
I personally would like the `_register` function to accept a dictionary instead of `*names`. Then don't need the separate steps of storing and registering. But I get the feeling you are not a fan of dictionaries. So if you prefer this, I am fine with it as well.

Reply:
It's not that I'm not a fan of dictionary. Actually, one could even store automatically parameters as arguments without dictionaries, like this 

```python
self._register('nvars', 'ndim', 'c', 'stencil_type', 'order', 'bc', readOnly=True, localVars=locals())
# => get the parameters values from locals() and store it directly as attribute in _register method
```
But I'm following one of Python's Zen : _"Explicit is better than implicit"_

By setting all parameters as attributes in the _register method, you hide the storing of parameters as object attribute, and never write it explicitly, _e.g_ `self.nvars = nvars`. From a developer point of view, once you are aware of this mechanism, it is indeed simpler to write and avoid this whole `self.blabla = blabla` multiple times.
But this is only for one person, and she does it only one time ! While for someone who want read the code, specifying in the constructor `self.thisPar = thisPar` (or something else) is explicit and easier to understand, especially if you are not aware of the `_register` mechanism.

Reply:
I'd rather have one mechanism of "registering" attributes than two (which have to be done both). If you are clear about what the `_register` method does, I think there will be no confusion. Maybe with a more explicit name for this method (e.g. `_make_attribute_and_register`) this is not such a problem? The current proposal is not clear to me (if I were someone who wants to read the code): why do I need to store AND register? And what happens if I store but don't register?

Reply:
@pancetta I do like you idea of making the method name more explicit. I've implemented this in the last commit. Also on a side note : paramters are stored in `set`, as looking up an element is faster than in a list. But then the order they are given in the `_makeAttributeAndRegister` function does not matter, only the alphabetic order is used, see :

![image](https://user-images.githubusercontent.com/12183839/212850547-ee1ffbf5-1638-4289-8d69-118f2a00f117.png)

Since setting attribute of the class method is (theoretically) done only during the instantiation with `__init__`, it should not degrade that much performance to use `list` instead, which would allow to keep the parameter order in the `params` dictionary. Would you prefer that ?

Reply:
Any scenario where the order would matter? If not, a `set` it is.